### ISSUE TITLE
Recent regression in checkver.ps1

### ISSUE THREAD

**Issue by nightroman** [lukesampson/scoop#1704]
The recent change in *checkver.ps1* causes some problems. It looks like the reason is using the PowerShell automatic variable `$matches` as the own local variable:

https://github.com/lukesampson/scoop/blob/82ec0fe7a11c1fc0db39cb51b671a5650c1252c8/bin/checkver.ps1#L150-L162

**Comment by nightroman** [comment]
Namely, I have a directory with .json manifests.

The command 

    C:\Users\...\AppData\Local\scoop\apps\scoop\current\bin\checkver.ps1 -dir .

used to work. Now it fails with:

````
Exception calling "Add" with "2" argument(s): "Key cannot be null.
Parameter name: key"
At C:\Users\...\AppData\Local\scoop\apps\scoop\current\bin\checkver.ps1:158 char:35
+ ...      $matches.groups | % { $matchesHashtable.Add($_.Name, $_.Value) }
+                                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
    + FullyQualifiedErrorId : ArgumentNullException
````

**Comment by nightroman** [comment]
Another potential cause is that `[regex]::matches` gets `MatchCollection`. This type does not have a property `groups`, but the code calls `$matches.groups`. Do I miss something obvious?

**Comment by nightroman** [comment]
Last known working fine commit is
8b41c5e ("Update unbound to version 1.6.5", 2017-08-21)

**Comment by r15ch13** [comment]
Which powershell version are you using?

**Comment by nightroman** [comment]
````
Name                           Value
----                           -----
PSVersion                      5.1.14394.1000
PSEdition                      Desktop
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
BuildVersion                   10.0.14394.1000
CLRVersion                     4.0.30319.42000
WSManStackVersion              3.0
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
````

**Comment by nightroman** [comment]
Regardless of the version in use, there are two issues:

- `$matches` is the PowerShell automatic variable. It should not be used for writing.
- `[regex]::matches` gets `MatchCollection`. This type has neither `Success`, nor `Groups` members. But the code uses them. This looks not quite right. Definitely not for the strict mode, for example.

**Comment by r15ch13** [comment]
You're right about the `$matches` variable. Will change that.

Regarding the `MatchCollection`: There are `select-object -last 1` and `select-object -first 1` are used on the result, which returns a single [`System.Text.RegularExpressions.Match`](https://msdn.microsoft.com/en-us/library/system.text.regularexpressions.match(v=vs.110).aspx).

**Comment by nightroman** [comment]
This is what I see in VSCode debugger for `$match.Groups`:
```
Groups   : {Dependency Walker 2.2, 2.2}
Success  : True
Captures : {Dependency Walker 2.2}
Index    : 1260
Length   : 21
Value    : Dependency Walker 2.2

Success  : True
Captures : {2.2}
Index    : 1278
Length   : 3
Value    : 2.2
```

**Comment by nightroman** [comment]
There is no property `Name`.

**Comment by nightroman** [comment]
Found it:

https://msdn.microsoft.com/en-us/library/system.text.regularexpressions.group.name
Version Information
.NET Framework
**Available since 4.7**

**Comment by r15ch13** [comment]
Hm, okay I didn't know this Property is only available since .NET 4.7
Do you know a workaround for this?
`checkver.ps1` was created mainly for the automated update service, which uses Powershell 6.

**Comment by nightroman** [comment]
AFAIR, *checkver.ps1* was created a long time ago as an informational tool.
I probably can help find a "workaround" if I know the goal.

**Comment by r15ch13** [comment]
It's used for creating named variables out of the named capture groups, so they can be used in the manifests autoupdate block. 
See [git.json](https://github.com/lukesampson/scoop/blob/master/bucket/git.json) for example (`$matchVersion` and `$matchShort` based on `v(?<version>[\\d\\w.]+)/PortableGit-(?<short>[\\d.]+).*\\.exe`).

All available substitutions get created in the [get_version_substitutions()](https://github.com/lukesampson/scoop/blob/master/lib/autoupdate.ps1#L200) function.

**Comment by nightroman** [comment]
https://stackoverflow.com/questions/1381097/how-do-i-get-the-name-of-captured-groups-in-a-c-sharp-regex

**Comment by r15ch13** [comment]
Fixed 😄

**Comment by nightroman** [comment]
It works, thank you!

**Comment by inosik** [comment]
I now have another issue with `checkver1.ps1`:

```
docfx: Unable to cast object of type 'System.Management.Automation.PSObject' to type 'System.Type'.
In C:\Users\nosikil\scoop\apps\scoop\current\bin\checkver.ps1:154 Zeichen:13
+             $match = $regex::matches($page, $regexp) | select-object -first 1
+             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : OperationStopped: (:) [], InvalidCastException
    + FullyQualifiedErrorId : System.InvalidCastException

couldn't match '\/releases\/tag\/(?:v)?([\d.]+)' in https://github.com/dotnet/docfx/releases/latest
```

I have PowerShell 3 on Windows 7:

```
Name                           Value
----                           -----
PSVersion                      3.0
WSManStackVersion              3.0
SerializationVersion           1.1.0.1
CLRVersion                     4.0.30319.42000
BuildVersion                   6.2.9200.22198
PSCompatibleVersions           {1.0, 2.0, 3.0}
PSRemotingProtocolVersion      2.2
```

**Comment by nightroman** [comment]
I think both `$regex::matches($page, $regexp)` should be replaced with `$regex.matches($page)` -- mind one argument and `.` instead of `::`.

**Comment by inosik** [comment]
@nightroman That was the fix! See #1707.
